### PR TITLE
More CRAM Encoding tests + updates

### PR DESCRIPTION
**Pull request by jmthibault79**
### Description

External*CodecTests
ByteArrayLenCodec test
ITF8 and LTF8 updates

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

File: src/main/java/htsjdk/samtools/cram/io/ITF8.java [hunk 1]
```diff
@@ -41,6 +69,48 @@ public static int readUnsignedITF8(final InputStream inputStream) throws IOExcep
         return ((b1 & 15) << 28) | inputStream.read() << 20 | inputStream.read() << 12 | inputStream.read() << 4 | (15 & inputStream.read());
     }
 
+    /**
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/io/ITF8.java]: Moved up to group the read methods together

File: src/main/java/htsjdk/samtools/cram/io/ITF8.java [hunk 2]
```diff
@@ -51,136 +121,106 @@ public static int readUnsignedITF8(final InputStream inputStream) throws IOExcep
...
-        return value;
-    }
-
-    /**
-     * Writes an unsigned (32 bit) integer to a byte new array encoded as ITF8. The sign bit is interpreted as a value bit.
-     *
-     * @param value the value to be written out
-     * @return the bytes holding ITF8 representation of the value
-     */
-    public static byte[] writeUnsignedITF8(final int value) {
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/io/ITF8.java]: moved down to group the write methods together

File: src/main/java/htsjdk/samtools/cram/io/ITF8.java [hunk 3]
```diff
@@ -51,136 +121,106 @@ public static int readUnsignedITF8(final InputStream inputStream) throws IOExcep
...
-    }
-
     /**
      * Writes an unsigned (32 bit) integer to an {@link OutputStream} encoded as ITF8. The sign bit is interpreted as a value bit.
      *
      * @param value the value to be written out
      * @param buffer   the {@link ByteBuffer} to write to
      */
-    public static void writeUnsignedITF8(final int value, final ByteBuffer buffer) {
+    public static int writeUnsignedITF8(final int value, final ByteBuffer buffer) {
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/io/ITF8.java]: easy to return the length, so why not

File: src/main/java/htsjdk/samtools/cram/structure/EncodingID.java [hunk 4]
```diff
@@ -46,7 +46,7 @@
      */
     BYTE_ARRAY_STOP,
     /**
-     * http://en.wikipedia.org/wiki/Beta_Code
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/EncodingID.java]: A different unrelated Beta Code

File: src/test/java/htsjdk/samtools/cram/encoding/ByteArrayLenCodecTest.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import org.testng.annotations.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ByteArrayLenCodecTest extends HtsjdkTest {
```
Comment (cmnbroad) [inline src/test/java/htsjdk/samtools/cram/encoding/ByteArrayLenCodecTest.java]: This test class is called `ByteArrayLenCodecTest`, but `ByteArrayLenCodec` is a private class inside of `ByteArrayLenEncoding`. That makes me think this should be called `ByteArrayLenEncodingTest. But I'm not sure if the tests should change, or the actual classes ? Is this is part of the weirdness you were talking about yesterday ?

File: src/test/java/htsjdk/samtools/cram/encoding/ByteArrayLenCodecTest.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            byteArrayOutputStream.write((byte) EncodingID.EXTERNAL.ordinal());
+            final byte[] byteBytes = new byte[]{valuesExternalBlockId};
+            ITF8.writeUnsignedITF8(byteBytes.length, byteArrayOutputStream);
+            byteArrayOutputStream.write(byteBytes);
+
+            ByteArrayLenEncoding bale = new ByteArrayLenEncoding();
+            bale.fromByteArray(byteArrayOutputStream.toByteArray());
+            return bale;
+        }
+    }
```
Comment (cmnbroad) [inline src/test/java/htsjdk/samtools/cram/encoding/ByteArrayLenCodecTest.java]: @jmthibault79 It seems like this code is very similar to the actual code in `ByteArrayLenEncoding` (but its slightly different ?) I wonder if we shouldn't refactor `ByteArrayLenEncoding` and `ByteArrayLenCodec` so we can test them directly instead ? Or is that what you're planning to do in the next round ? It almost seems like we should wait for that if its the latter.

File: src/main/java/htsjdk/samtools/cram/io/ITF8.java [hunk 7]
```diff
@@ -8,9 +8,37 @@
...
+ *      write out [bits 13-20]
+ *      write out [bits 5-12]
+ *      write out [bits 1-8]
+ *
  */
 public class ITF8 {
 
+    public static final int MAX_BYTES = 5;
+    public static final int MAX_BITS = 8 * MAX_BYTES;
+
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/io/ITF8.java]: They *could* be useful outside of this?  But I have no specific use case, so I can make them pkg-priv

File: src/main/java/htsjdk/samtools/cram/io/ITF8.java [hunk 8]
```diff
@@ -150,37 +174,62 @@ public static int readUnsignedITF8(final ByteBuffer buffer) {
...
+    }
+
+    /**
+     * Writes an unsigned (32 bit) integer to a byte new array encoded as ITF8. The sign bit is interpreted as a value bit.
+     *
+     * @param value the value to be written out
+     * @return the bytes holding ITF8 representation of the value
+     */
+    public static byte[] writeUnsignedITF8(final int value) {
+        final ByteBuffer buffer = ByteBuffer.allocate(10);
```
Comment (cmnbroad) [inline src/main/java/htsjdk/samtools/cram/io/ITF8.java]: I know this code was just moved from elsewhere, but is 10 meaningful here ? Although harmless, it seems like random over-allocation by half. Is there any way possible this can use more than the max possible # of bytes in an ITF8 encoding for an int, which is 5?. Can this use the new `MAX_BYTES` constant ?

File: src/test/java/htsjdk/samtools/cram/io/IOTestCases.java [hunk 9]
```diff
@@ -28,15 +28,74 @@
...
+        list.add((byte)-1);
+
+        // scan with bits:
+        for (int i = 0; i < 7; i++) {
+            list.add((byte)((1 << i) - 2));
+            list.add((byte)((1 << i) - 1));
+            list.add((byte)(1 << i));
+            list.add((byte)(-(1 << i) + 2));
+            list.add((byte)(-(1 << i) + 1));
+            list.add((byte)-(1 << i));
```
Comment (cmnbroad) [inline src/test/java/htsjdk/samtools/cram/io/IOTestCases.java]: Nice!

File: src/main/java/htsjdk/samtools/cram/io/CramIntArray.java [hunk 10]
```diff
@@ -15,7 +15,7 @@
      * @return array of integers from the input stream
      * @throws IOException as per java IO contract
      */
-    public static int[] array(final InputStream inputStream) throws IOException {
+    public static int[] array(final InputStream inputStream) {
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/io/CramIntArray.java]: good catch, thanks

File: src/main/java/htsjdk/samtools/cram/io/ITF8.java (hunk 8 above)
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/io/ITF8.java]: That makes sense to me.  I'll see how the tests respond.

File: src/test/java/htsjdk/samtools/cram/encoding/ExternalByteArrayCodecTest.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.io.*;
+
+public class ExternalByteArrayCodecTest extends HtsjdkTest {
+
+    @Test(dataProvider = "testByteArrays", dataProviderClass = IOTestCases.class)
+    public void codecTest(final byte[] values) throws IOException {
+
+        try (final ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+            final BitCodec<byte[]> writeCodec = new ExternalByteArrayCodec(os, null);
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/test/java/htsjdk/samtools/cram/encoding/ExternalByteArrayCodecTest.java]: Yes, exactly.

Separating these is one of the goals of an upcoming refactor sweep.
